File: lightdarktest.m
function lightdarktest(mode,outfile)
% code to perform light and dark test
% mode = 0 (dark test) or 1 (light test)
% saves data to outfile

PTBprelims

switch mode
    case 0
        task = 'darktest';
        numtrials = 3; %number of light/dark cycles
        habit_mat = 255 * ones(1,3); %habituate to full bright
        stim_mat = zeros(numtrials,3); %dark during stimulus
        rec_mat = 255*ones(numtrials,3); %full bright during recovery
        stim_time = 1; %duration of the dark in seconds
    case 1
        task = 'lighttest';
        stim_mat = 255 * [ [0.25 0.25 0.25] ; %bright during stimulus
            [0.5 0.5 0.5];
            [0.75 0.75 0.75];
            [1 1 1] ] ;
        numtrials=size(stim_mat,1);
        rec_mat = zeros(numtrials,3); %dark during recovery
        habit_mat = zeros(1,3); %habituate to full bright
        stim_time = 0.2;
end

habituation_dur = 10; %habituation time (in s) before first flash
recover_dur = 8*ones(1,numtrials); %recovery time post-flash (in s)
stim_dur = stim_time*ones(1,numtrials);  %duration of each stimulus (in s)

%display onscreen countdown
countdown

% *************************************************************************
%
% Start tracking and plot the gaze data read from the tracker.
%
% *************************************************************************

%put on habituation stimulus
Screen('FillRect',win,habit_mat,[]);
Screen('Flip',win);
WaitSecs(habituation_dur);

for ind=1:numtrials
    
    tetio_startTracking;
    
    WaitSecs(2);
    
    %paint light stimulus onscreen
    Screen('FillRect',win,stim_mat(ind,:),[]);
    Screen('Flip',win);
    
    %Record Time of Stim. Onset
    %StimOnSet(ind)=GetSecs;
    %not sure about the syncing of time so alternatively:
    data(ind).ontime=tetio_localToRemoteTime(tetio_localTimeNow());
    
    %wait the duration of the stimulus
    WaitSecs(stim_dur(ind));
    
    %clear stimulus
    Screen('FillRect',win,rec_mat(ind,:));
    Screen('Flip',win);
   
    %Record Time Stimulus goes off
    %StimOff(ind)=GetSecs;
    data(ind).offtime=tetio_localToRemoteTime(tetio_localTimeNow());
    
    %wait recovery time
    WaitSecs(recover_dur(ind));
    
    tetio_stopTracking;
    
    %read eye data
    [lefteye, righteye, timestamp, trigSignal] = tetio_readGazeData;
    
    %%%% save data each trial %%%%%%
    data(ind).lefteye = lefteye;
    data(ind).righteye = righteye;
    data(ind).timestamp = timestamp;
    data(ind).trig = trigSignal;
    save(outfile,'data','task')
    
end



%tetio_cleanUp; %%%% are we sure we need to do this? not really...

% return the screen to dark
Screen('FillRect',win,[0 0 0],[]);
Screen('Flip',win);

disp('Program finished.');

File: pytask/lightdarktest.py
'''
Code for light/dark test orginally coded in Matlab.
'''
from psychopy import visual, core
import numpy as np
import display

# mode = 0 (dark test) or 1 (light test), outfile is the file that data is
# saved to


def lightdarktest(controller, mode, outfile):
    # Create window to display test
    testWin = controller.testWin
    if mode == 0:  # dark test
        task = 'darktest'
        numtrials = 3  # number of light/dark cycles
        habit_mat = [1, 1, 1]  # set habituation to full brightness
        # set stimulus matrix to darkness
        stim_mat = np.negative(np.ones((numtrials, 3)))
        # back to full brightness for recovery
        rec_mat = np.ones((numtrials, 3))
        stim_time = 1  # duration of stimulus in seconds
        display.text(controller.experWin, 'Running Dark Test')
    elif mode == 1:  # light test
        task = 'lighttest'
        habit_mat = [-1, -1, -1]  # habituation is black
        stim_mat = [[-0.5, -0.5, -0.5],  # stimulus is increasing brightness each time
                    [0.00, 0.00, 0.00],
                    [0.50, 0.50, 0.50],
                    [1.00, 1.00, 1.00]]
        # set numtrials based on number of stimulus conditions
        numtrials = len(stim_mat)
        # recovery goes back to black
        rec_mat = np.negative(np.ones((numtrials, 3)))
        stim_time = 0.2  # length of stimulus time in seconds
        display.text(controller.experWin, 'Running Light Test')
    habit_dur = 10  # time for habituation
    # time for recovery between stimuli
    recover_dur = 8 * np.ones((1, numtrials))
    stim_dur = stim_time * np.ones((1, numtrials))  # duration of each stimulus
    display.countdown(controller)  # display countdown before task

    # habituation
    display.fill_screen(testWin, habit_mat)
    core.wait(habit_dur)

    # Start eye tracking
    if not controller.testing:
        controller.tobii_cont.setDataFile(outfile)
        controller.tobii_cont.startTracking()
        controller.tobii_cont.setEventsAndParams(['ontime', 'offtime', 'task'])
        controller.tobii_cont.setParam('task', task)

    for i in range(numtrials):
        core.wait(2.0)

        # display stimulus
        display.fill_screen(testWin, stim_mat[i])

        # record timestamp on tracker for start of stimulus
        if not controller.testing:
            controller.tobii_cont.recordEvent('ontime')

        # wait for stimulus
        core.wait(stim_dur[0][i])

        # display recovery
        display.fill_screen(testWin, rec_mat[i])

        # record timestamp on tracker for end of stimulus
        if not controller.testing:
            controller.tobii_cont.recordEvent('offtime')

        # wait for recovery
        core.wait(recover_dur[0][i])

    # End eye tracking
    if not controller.testing:
        controller.tobii_cont.stopTracking()
        controller.tobii_cont.closeDataFile()


if __name__ == '__main__':
    lightdarktest(0, None)
